R/join_keys-c.R

```diff
@@ -10,7 +10,7 @@
 #'   jk,
 #'   join_keys(
 #'     join_key("ds4", keys = c("pk4", "pk4_2")),
-#'     join_key("ds4", "ds3", c(pk4_2 = "pk3"))
+#'     join_key("ds3", "ds4", c(pk3 = "pk4_2"))
 #'   )
 #' )
 c.join_keys <- function(...) {
```

Comment:
@averissimo Here's a shorter version for c.join_keys. WDYT?

```
c.join_keys <- function(...) {
  args <- rlang::list2(...)  #Combined list2 calls for different subsets to reduce overhead.
  checkmate::assert_multi_class(args[[1]], classes = c("join_keys", "join_key_set"))
# Removed assert_list check for additional arguments as they're processed in Reduce function.

# Using the reduce function on the args list avoids unnecessary steps of separating and merging the first argument, simplifying the process.

  out <- Reduce(function(.x, .y) {
    # or use checkmate::assert_list()
    checkmate::assert_multi_class(.y, classes = c("join_keys", "join_key_set"))

    # Perform the merge operation
    merged <- utils::modifyList(.x, .y, keep.null = FALSE)
    parents(merged) <- .merge_parents(.x, .y)
    merged
  }, args, init = join_keys())

  out
}
```

Reply:
The reason why we are breaking the first argument from the rest is refered in the [comment above](https://github.com/insightsengineering/teal.data/pull/215/files#r1426659206).

_To sum it up_: we allow a unrestrictive merge between argument 1 and "the rest". The rest must not have any conflicting keys between themselves.

Short example of this in play can be read in the [same comment above](https://github.com/insightsengineering/teal.data/pull/215/files#r1426659206)

🔴 If this behavior is not necessary, them we can simplify it.

Reply:
note: the assertions are all in the top-level function to allow for more informative error messages, otherwise it would show `Error in f(init, x[[i]])` as it's define inside the Reduce function

Toy example:

``` r
and <- function(...) {
  data <- rlang::list2(...) |> purrr::flatten() |> unlist()
  checkmate::assert_logical(data)
  
  Reduce(
    f = \(.x, .y) {
      checkmate::assert_multi_class(.y, classes = c("join_keys", "join_key_set"))
      .x && .y
    },
    x = data,
    init = TRUE
  )
}

and(TRUE, 1)
#> Error in and(TRUE, 1):
#>   Assertion on 'data' failed: Must be of type 'logical', not 'double'.
and(TRUE, FALSE)
#> Error in f(init, x[[i]]):
#>   Assertion on '.y' failed: Must inherit from class 'join_keys'/'join_key_set', but has class 'logical'.
```

Reply:
>  If this behavior is not necessary, them we can simplify it.

Concatenated operation overriding the relationship when there are two datasets e.g in  `case1` and `case2,` but it throws an error in `case3` which is a bit odd. This may create confusion, so simplifying it and not restricting this behavior unlike join_keys() constructor might be better approach IMO.

Another argument for this is that we should not put a safeguard for everything, and this might be an intentional operation from the app-dev side.

```
##case1:################################
c(
  join_keys(join_key("ds1", "ds1", "original")),
  join_keys(
    join_key("ds1", "ds1", "first_change")
  )
) |> invisible()

##case2:################################
c(c(
  join_keys(join_key("ds1", "ds1", "original")),
  join_keys(
    join_key("ds1", "ds1", "first_change")
  )),
  join_keys(
    join_key("ds1", "ds1", "second change")
  )
) 

##case3:################################
c(
  join_keys(join_key("ds1", "ds1", "original")),
  join_keys(
    join_key("ds1", "ds1", "first_change")
  ),
  join_keys(
    join_key("ds1", "ds1", "second change")
  )
) |> invisible()
#> Error in stop_message(dataset_1_one, dataset_2_one):
#>   cannot specify multiple different join keys between datasets: ds1 and ds1
```

Reply:
> note: the assertions are all in the top-level function to allow for more informative error messages, otherwise it would show Error in f(init, x[[i]]) as it's define inside the Reduce function

we can add meaning full message in assertion there